etc/migration/migrateUsersToShibboleth.js

```diff
@@ -0,0 +1,352 @@
+ * @param  {Object}     callback.err            An error that occurred, if any
+ * @param  {Object[]}   callback.mappedUsers    The Shibboleth users that were mapped to existing users
+ * @api private
+ */
+var _mapUsersToShibLogin = function(allUsers, callback) {
+    // Remove users with duplicate emails or invalid login IDs
+    var cleanedUsers = _.chain(allUsers)
+        .filter(function(user) {
+            var valid = user.loginId;
+            if (!valid) {
+                _writeErrorRow(user, 'This user has an invalid loginId');
+            }
+            return valid;
+        })
+        .groupBy(function(user) {
+            return user.email;
+        })
+        .map(function(grouped) {
+            if (grouped.length === 1) {
+                return grouped[0];
+            } else {
+                _.each(grouped, function(dupe) {
+                    _writeErrorRow(dupe, 'Found more than one user with this email');
+                });
+            }
+        })
+        .compact()
+        .value();
+
+    if (_.isEmpty(cleanedUsers)) {
+        log().info('No suitable users found for tenant %s', tenantAlias);
+        callback(null, null);
+    }
+
+    _createShibLoginRecords(cleanedUsers, function(err, mappedUsers) {
+        if (err) {
+            callback(err);
+        }
+
+        callback(null, mappedUsers);
+    });
+};
+
+/**
+ * Create new authentication records linked to Shibboleth EPPN for all users within the tenant
+ *
+ * @param  {Object[]}   users                   The users to map
+ * @param  {Function}   callback                Standard callback function
+ * @param  {Object}     callback.err            An error that occurred, if any
+ * @param  {Object[]}   callback.mappedUsers    The Shibboleth users that were mapped to existing users
+ * @api private
+ */
+var _createShibLoginRecords = function(users, callback, _mappedUsers) {
+    _mappedUsers = _mappedUsers || [];
+
+    if (_.isEmpty(users)) {
+        return callback(null, _mappedUsers);
+    }
+
+    var userHash = users.shift();
+
+    log().info('Processing user %s', userHash.email);
+
+    // Create a new login record for user
+    _createNewUserLogin(userHash, function(err) {
+        if (err) {
+            callback(err);
+        }
+
+        _mappedUsers.push(userHash);
+        _createShibLoginRecords(users, callback, _mappedUsers);
+    });
+};
+
+/**
+ * Create new Shibboleth login details for existing account
+ *
+ * @param  {Object}     userHash                The user account we want to keep
+ * @param  {Function}   callback                Standard callback function
+ * @param  {Object}     callback.err            An error that occurred, if any
+ * @api private
+ */
+var _createNewUserLogin = function(userHash, callback) {
+    var email = userHash.email;
+    var userId = userHash.principalId;
+    var newLoginId = util.format('%s:shibboleth:%s', tenantAlias, email);
```

Comment:
A user can change their profile email to something that is no longer the original eppn. Since we want google login ids only to be remapped as shibboleth ids with the same external id, we can directly parse the email address portion of the google login ids and add a new entry with the `cam:shibboleth:` prefix.

So if I logged in with gmail, I would have a login id: `cam:google:mrvisser@cam.ac.uk`. When doing the mapping, we can parse out the `mrvisser@cam.ac.uk` and then add a new entry `cam:shibboleth:mrvisser@cam.ac.uk`.